File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head==NULL || head->next == NULL){
            return head;
        }
        ListNode* temp = new ListNode();
        ListNode* front = new ListNode();
        ListNode* prev = new ListNode();
        temp = head;
        front = head->next;
        prev = NULL;
        while(temp->next!=NULL){
            temp->next = prev;
            prev = temp;
            temp = front;
            front = front->next;
        }
        temp->next = prev;
        head = temp;
        return head;
    }
};

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null){
            return head;
        }

        ListNode front = head.next;
        ListNode curr = head;
        ListNode prev = null;
        while(curr!=null){
            curr.next = prev;
            prev = curr;
            curr = front;
            if(front!=null){
                front = front.next;
            }
        }
        return prev;

    }
}